File: leetcode/0269-alien-dictionary/solution.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>
using namespace std;

class Solution {
public:
    string alienOrder(vector<string>& words) {
        unordered_map<char, unordered_set<char>> graph;
        unordered_map<char, int> inDegree;

        for (const string& word : words) {
            for (char c : word) {
                graph[c];
                inDegree[c] = inDegree.count(c) ? inDegree[c] : 0;
            }
        }

        for (size_t i = 0; i + 1 < words.size(); i++) {
            const string& w1 = words[i];
            const string& w2 = words[i + 1];
            size_t minLen = min(w1.size(), w2.size());
            bool foundDiff = false;

            for (size_t j = 0; j < minLen; j++) {
                if (w1[j] != w2[j]) {
                    if (!graph[w1[j]].count(w2[j])) {
                        graph[w1[j]].insert(w2[j]);
                        inDegree[w2[j]]++;
                    }
                    foundDiff = true;
                    break;
                }
            }

            if (!foundDiff && w1.size() > w2.size()) {
                return "";
            }
        }

        queue<char> q;
        for (auto& [c, deg] : inDegree) {
            if (deg == 0) {
                q.push(c);
            }
        }

        string order;
        while (!q.empty()) {
            char c = q.front();
            q.pop();
            order += c;
            for (char next : graph[c]) {
                if (--inDegree[next] == 0) {
                    q.push(next);
                }
            }
        }

        return order.size() == inDegree.size() ? order : "";
    }
};

File: leetcode/0269-alien-dictionary/Solution.java
import java.util.*;

class Solution {
    public String alienOrder(String[] words) {
        Map<Character, Set<Character>> graph = new HashMap<>();
        Map<Character, Integer> inDegree = new HashMap<>();

        for (String word : words) {
            for (char c : word.toCharArray()) {
                graph.putIfAbsent(c, new HashSet<>());
                inDegree.putIfAbsent(c, 0);
            }
        }

        for (int i = 0; i < words.length - 1; i++) {
            String w1 = words[i], w2 = words[i + 1];
            int minLen = Math.min(w1.length(), w2.length());
            boolean foundDiff = false;

            for (int j = 0; j < minLen; j++) {
                char c1 = w1.charAt(j), c2 = w2.charAt(j);
                if (c1 != c2) {
                    if (!graph.get(c1).contains(c2)) {
                        graph.get(c1).add(c2);
                        inDegree.put(c2, inDegree.get(c2) + 1);
                    }
                    foundDiff = true;
                    break;
                }
            }

            if (!foundDiff && w1.length() > w2.length()) {
                return "";
            }
        }

        Queue<Character> queue = new LinkedList<>();
        for (char c : inDegree.keySet()) {
            if (inDegree.get(c) == 0) {
                queue.add(c);
            }
        }

        StringBuilder order = new StringBuilder();
        while (!queue.isEmpty()) {
            char c = queue.poll();
            order.append(c);
            for (char next : graph.get(c)) {
                inDegree.put(next, inDegree.get(next) - 1);
                if (inDegree.get(next) == 0) {
                    queue.add(next);
                }
            }
        }

        return order.length() == inDegree.size() ? order.toString() : "";
    }
}
